subprojects/execution/src/main/java/org/gradle/internal/execution/caching/impl/DefaultCachingStateBuilder.java

```diff
@@ -58,10 +54,7 @@ public final void withImplementation(ImplementationSnapshot implementation) {
     @OverridingMethodsMustInvokeSuper
     protected void processImplementation(ImplementationSnapshot implementation) {
         if (implementation.isUnknown()) {
-            noCachingReasonsBuilder.add(new CachingDisabledReason(
-                NON_CACHEABLE_IMPLEMENTATION,
-                "Implementation type " + implementation.getUnknownReason()
-            ));
+            throw new GradleException("Cannot calculate build cache key for work item with unknown implementation");
```

Comment:
Is there a more specific exception we could/should throw here?

Where does this message appear? I don't see it in any of the tests.

Reply:
This exception never should be thrown, since validation fails in the case and then we won't calculate the build cache key. Shall we remove these checks altogether?

Reply:
If it's an assertion, then we should throw `AssertionError`. But it would be a lot better to avoid this problem completely. I wonder if we can remove tracking unknown implementations. If not, then just not checking here is also a good alternative.

Reply:
I think we can't yet remove tracking unknown implementations, given that we still need to store them in the task history. As soon as we fail when we encounter an unknown implementation, I think we can stop tracking them.

Reply:
Since we don't track the individual reasons now for the implementation to be unknown, can we still ditch `ImplementationSnapshot.isUnknown()` and just pass around `null` or something?

Reply:
We need the reason for the validation, since this is what we'll print out. So we can't just use `null` as implementation snapshot.